### PR TITLE
Add "transformRequest" method to fetcher options

### PR DESCRIPTION
**Pull request by danawoodman**
Add support for a `transformRequest` method that takes a `RequestLike` (basically `RequestInit` with required fields and the `url`) and allows for mutating the request before it gets sent off. This allows for things like:

- Appending a query parameter or header to every request
- Modifying the method (not sure why you'd want to, but you _can_)
- Do some global mutation on the request body, like adding a required attribute or something
- Add custom behavior for a special route (eg only add request header to `/admin`)
- Replacing the base URL with something entirely if running in test mode/development

### Examples

```ts
// Add a header to every request
fetcher({
  transformRequest(req) {
    req.headers['Authorization'] = token
    return req
  }
})
```

```ts
// Add a query param to every request
fetcher({
  transformRequest(req) {
    const url = new URL(req.url)
    url.searchParams.set('message', 'hello world')
    req.url = url.toString()
    return req
  },
})
```

```ts
// Only add a header on a "/admin" route
fetcher({
  transformRequest(req) {
    const url = new URL(req.url)
    if (url.pathname.startsWith('/admin')) {
      req.headers['Authorization'] = token
    }
    return req
  },
})
```

```ts
// Change the origin of a URL only in dev mode
fetcher({
  transformRequest(req) {
    if (dev) {
      req.url = req.url.replace('prod.example.com', 'dev.example.com')
    }
    return req
  }
})
```

```ts
// Only apply a token on a specific domain
fetcher({
  transformRequest(req) {
    if (req.url.startsWith("https://api.example.com") {
      req.headers["Authorization"] = `Bearer ${token}`
    }
    return req
  }
})
```

```ts
// Util function to inject a user auth token into a request
function api_client(token: string) {
  return fetcher({
    base: "https://api.example.com",
    transformRequest(req) {
      req.headers['Authorization'] = token
      return req
    }
  })
}

// Somewhere else...
const token = read_jwt(jwt_cookie)
const api = api_client(token)
api.get("/foo")
```

### Additional notes

- Simplified the typing of FetcherOptions
- Exporting some types as it can be helpful both in tests but also when consuming, so you can conform to a type (eg the type of options for example)
- Simplify dev script

### TODO

- [ ] Add more tests (of what??)
- [x] Document this feature
- [ ] Optimize for file size/bytes

### CODE REVIEW

File: src/itty-fetcher.ts [hunk 1]
```diff
@@ -46,24 +49,40 @@ const fetchy =
...
       body: JSON.stringify(payload),
-    }).then((response) => {
+    }
+
+    /**
+     * Transform the outgoing request if a transformRequest function is provided
+     * in the options. This allows the user to modify the request before it is
+     * sent.
+     */
+    if (options.transformRequest) req = options.transformRequest(req)
```
Comment (danawoodman) [inline src/itty-fetcher.ts:81]:
```TS
if (typeof options?.transformRequest === 'function') {
```

Might want to check the type is a function?

Also, do we want to verify the return value is correct or do we just let things explode?

File: src/itty-fetcher.spec.ts [hunk 2]
```diff
@@ -25,7 +27,105 @@ describe('fetcher', () => {
       })
   })
 
-  const defaults = fetcher()
+  describe('transformRequest', () => {
```
Comment (kwhitley) [inline src/itty-fetcher.spec.ts]: only comment is:

- could you please move this to inside config options, under autoParse I guess? it's def part of the options being tested :)

File: src/itty-fetcher.ts (hunk 1 above)
Comment (kwhitley) [reply src/itty-fetcher.ts:81]: Not sure... I'm pretty down with fewer guard rails, allowing things to stay small (by requiring some self-discipline)

Comment (kwhitley) [reply src/itty-fetcher.ts:81]: I [usually] leave beautiful error messaging and checks for much bigger frameworks... :D

**Review by kwhitley — APPROVED** [review]

**Pull request merged**